src: port coverage serialization to C++

joyeecheung opened the pull request:
This patch moves the serialization of coverage profiles into
C++. With this we no longer need to patch `process.reallyExit`
and hook into the exit events, but instead hook into relevant
places in C++ which are safe from user manipulation. This also
makes the code easier to reuse for other types of profiles.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/21818/

bcoe commented:
@joyeecheung fascinating, I'll provide a thorough code review sometime tomorrow (including checking out the branch and testing). Assuming that collecting coverage continues to work as expected for the Node.js codebase and user-land modules, this refactor seems like a no-brainer.

Great work.

joyeecheung commented:
I updated this patch to accommodate https://github.com/nodejs/node/pull/26544 - we now must use `env->env_vars()` to query `NODE_V8_COVERAGE` for workers. In addition I removed the inheritance of `BaseObject` from `V8InspectorConnection`, since now the entire functionality is implemented in C++, we can just manage the lifetime of the connections along with Environment.

cc @addaleax

joyeecheung commented:
Ping @bcoe have you gotten around to this?

also trying to get more reviews... @nodejs/v8-inspector @nodejs/process @nodejs/fs

ofrobots commented:
@bcoe The OOM is probably related to the heap V8 has allocated rather than physical memory size. Try a bigger heap e.g. `--max-old-space-size=1024`.

joyeecheung commented:
> are you bumping into this issue with make coverage; I can try on my home computer as well (It's worth noting this also happened to me on master).

I ran into this before, try cleaning up the `.coverage` your build directory - if you run the command multiple times, it tries to combine all the results from previous runs so if you don't clean them up, the coverage data will grow forever.

joyeecheung commented on src/env.h:
```diff
@@ -1060,6 +1065,15 @@ class Environment {
+      std::unique_ptr<profiler::V8CoverageConnection> connection);
+  profiler::V8CoverageConnection* coverage_connection();
+
+  inline void set_coverage_directory(const char* directory);
```
This is set as usual during pre-execution, so yes it is resolved.

joyeecheung commented:
Landed in 864860e

addaleax commented on src/node_process_methods.cc:
```diff
@@ -171,6 +171,12 @@ static void Kill(const FunctionCallbackInfo<Value>& args) {
+    // TODO(joyeecheung): white list the signals?
+
+#if HAVE_INSPECTOR
+  profiler::EndStartedProfilers(env);
```
@joyeecheung Should this be guarded to only be run if `sig > 0` and `pid` is one of `0, -1, getpid(), -getpid()` or something similar?

joyeecheung replied:
I think we should only write the profiles when we are sure the kill is initiated to terminate the process, are those conditions enough for us to be sure about that?

addaleax replied:
@joyeecheung I mean, some signals that would usually terminate the process can be caught (like `SIGINT`), so we cannot be sure of whether this is going to terminate the process or not no matter what…

joyeecheung replied:
Maybe we can only take care of signals that don't have a handler attached through Node.js's APIs, though even then it would get complicated because of workers. Or we could just not guard this at all, or let the user decide somehow, but then no one has reported any issues with this so I don't know what the real-world use cases would be.
